### PR TITLE
EZP-29089: Usability bug when deleting containers

### PR DESCRIPTION
**Pull request by ViniTou**
| Question      | Answer
| ------------- | ---
| Tickets       | https://jira.ez.no/browse/EZP-29089
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Tests pass?   | yes
| Doc needed?   | no
| License       | [GPL-2.0](https://github.com/ezsystems/ezplatform-admin-ui/blob/master/LICENSE)

## Description

This checks if location is non-empty container and if it is displays additional info modal. If content have asset fieldtype additional modals are displayed after first one.

#### Checklist:
- [x] Coding standards (`$ composer fix-cs`)
- [x] Ready for Code Review

### CODE REVIEW

File: src/lib/Specification/Location/HasChildren.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    /**
+     * @param \eZ\Publish\API\Repository\Values\Content\Location $item
+     *
+     * @return bool
+     */
+    public function isSatisfiedBy($item): bool
+    {
+        $children = $this->locationService->loadLocationChildren($item);
+
+        return 0 !== $children->totalCount;
```
Comment (mikadamczyk) [inline src/lib/Specification/Location/HasChildren.php]: maybe `0 < $children->totalCount`

File: src/bundle/Resources/public/js/scripts/admin.trash.container.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+(function (global, doc) {
+    const toggleForm = doc.querySelector('form[name="location_trash_container"]');
+    const haveAsset = toggleForm.dataset.haveAsset;
```
Comment (sunpietro) [inline src/bundle/Resources/public/js/scripts/admin.trash.container.js]: `hasAsset`

File: src/bundle/Resources/public/js/scripts/admin.trash.container.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    const haveUniqueAsset = toggleForm.dataset.haveUniqueAsset;
```
Comment (sunpietro) [inline src/bundle/Resources/public/js/scripts/admin.trash.container.js]: `hasUniqueAsset`

File: src/bundle/Resources/public/js/scripts/admin.trash.container.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+    const openTrashImageAssetModal = (event) => {
+        if (!haveAsset && !haveUniqueAsset) {
+            return;
+        }
+        event.preventDefault();
```
Comment (sunpietro) [inline src/bundle/Resources/public/js/scripts/admin.trash.container.js]: CS: missing spaces before and after `event.preventDefault();`

File: src/bundle/Resources/public/js/scripts/button.state.checkbox.toggle.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
+(function (global, doc) {
+    const toggleForms = [...doc.querySelectorAll('.ez-toggle-btn-state-checkbox')];
```
Comment (sunpietro) [inline src/bundle/Resources/public/js/scripts/button.state.checkbox.toggle.js]: Due to latest changes in the browsers specification implementation, `doc.querySelectorAll('.ez-toggle-btn-state-checkbox')` is just enough. No need to convert it to an array to use with `forEach` only.

File: src/bundle/Resources/public/js/scripts/button.state.checkbox.toggle.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    const ALL_CHECKED = 'all-checked';
+    const ANY_CHECKED = 'any-checked';
+
+    toggleForms.forEach((toggleForm) => {
+        const checkboxInputs = [...toggleForm.querySelectorAll('input[type="checkbox"]')];
```
Comment (sunpietro) [inline src/bundle/Resources/public/js/scripts/button.state.checkbox.toggle.js]: Same as above.

File: src/bundle/Resources/public/js/scripts/button.state.checkbox.toggle.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+        const toggleButtonState = () => {
```
Comment (sunpietro) [inline src/bundle/Resources/public/js/scripts/button.state.checkbox.toggle.js]: This function should be implemented outside of `forEach` loop. If you want to have an access to the `checkboxInputs` then a useful piece of information is that a `forEach` loop callback takes 3 params: item, index, an iterated array with all values. See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach

File: src/bundle/Resources/public/js/scripts/button.state.checkbox.toggle.js [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+            const button = doc.querySelector(toggleForm.dataset.toggleButtonId);
+
+            const anyIsChecked = checkboxInputs.some(el => el.checked);
```
Comment (sunpietro) [inline src/bundle/Resources/public/js/scripts/button.state.checkbox.toggle.js]: CS: all variables should come one after another without additional empty lines between them.

File: src/bundle/Resources/views/content/modal_location_trash_container.html.twig [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+                        'class' : 'ez-toggle-btn-state-checkbox',
+                        'data-toggle-button-id' : '#location_trash_container_trash',
+                        'data-has-asset': has_asset,
+                        'data-has-unique-asset': has_unique_asset
+                    },
+                }) }}
+                {{ form_widget(form.trashContainer) }}
+            </div>
+            <div class="modal-footer">
+                <button type="button" class="btn btn-secondary btn--no" data-dismiss="modal">
```
Comment (tischsoic) [inline src/bundle/Resources/views/content/modal_location_trash_container.html.twig]: According to our style guidelines (https://doc.ezplatform.com/en/2.2/guidelines/components/buttons/) this button probably should have classes `btn btn-dark`.

File: src/bundle/Resources/public/js/scripts/button.state.checkbox.toggle.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
+(function (global, doc) {
+    const toggleForms = doc.querySelectorAll('.ez-toggle-btn-state-checkbox');
+    const ALL_CHECKED = 'all-checked';
+    const ANY_CHECKED = 'any-checked';
+
+    const toggleButtonState = (toggleForm, checkboxInputs) => {
+        const button = doc.querySelector(toggleForm.dataset.toggleButtonId);
+        const anyIsChecked = checkboxInputs.some(el => el.checked);
```
Comment (sunpietro) [inline src/bundle/Resources/public/js/scripts/button.state.checkbox.toggle.js]: `isAnyChecked`

File: src/bundle/Resources/public/js/scripts/button.state.checkbox.toggle.js [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+        }
+
+        button[methodName]('disabled', true);
+    };
+
+    toggleForms.forEach((toggleForm) => {
+        const checkboxInputs = toggleForm.querySelectorAll('input[type="checkbox"]');
+
+        checkboxInputs.forEach((input, index, inputs) => {
+            input.addEventListener('change', toggleButtonState.bind(null, toggleForm, [...inputs]), false);
```
Comment (sunpietro) [inline src/bundle/Resources/public/js/scripts/button.state.checkbox.toggle.js]: Do you need `[...inputs]`?

File: src/bundle/Resources/public/js/scripts/button.state.checkbox.toggle.js [hunk 12]
```diff
@@ -0,0 +1,27 @@
+(function (global, doc) {
+    const toggleForms = doc.querySelectorAll('.ez-toggle-btn-state-checkbox');
+    const ALL_CHECKED = 'all-checked';
+    const ANY_CHECKED = 'any-checked';
+
+    const toggleButtonState = (toggleForm, checkboxInputs) => {
+        const button = doc.querySelector(toggleForm.dataset.toggleButtonId);
+        const anyIsChecked = checkboxInputs.some(el => el.checked);
+        const allChecked = checkboxInputs.every(el => el.checked);
```
Comment (sunpietro) [inline src/bundle/Resources/public/js/scripts/button.state.checkbox.toggle.js]: I think this line and the line above can be moved into the line line no. 22 and 23, for better performance.

File: src/bundle/Resources/public/js/scripts/button.state.checkbox.toggle.js (hunk 11 above)
Comment (ViniTou) [reply src/bundle/Resources/public/js/scripts/button.state.checkbox.toggle.js]: I can't use `.some` and `.every` on Node Collection, unless there is some other way to do this.

**Comment by lserwatka** [comment]
@mnocon Behat is failing here

**Comment by lserwatka** [comment]
Perhaps rebase is needed with master after your fix will be merged up.

**Comment by lserwatka** [comment]
@ViniTou could you rebase with master?

**Pull request merged**